### PR TITLE
Bug fix/rlp 1097/batching get logs

### PR DESCRIPTION
**Pull request by JONAF2103**
### What was wrong?

RSK doesn't allow `eth_getLogs` calls with a big block size number. The fromBlock and toBlock params have to be near because the RSKj node doesn't suppport big queries.

### How was it fixed?

To fix this we added support to make small batch queries instead of making a huge one. Also we added some cache to store the logs to avoid having to query them all the time.

### CODE REVIEW

File: web3/eth.py [hunk 1]
```diff
@@ -350,8 +356,10 @@ def getFilterLogs(self, filter_id):
         )
 
     def getLogs(self, filter_params):
-        return self.web3.manager.request_blocking(
-            "eth_getLogs", [filter_params],
+        return self.logs_manager.get_logs(
+            filter_params,
+            self.getBlock('earliest')["number"],
+            self.getBlock('latest')["number"]
```
Comment (mortelli) [inline web3/eth.py:361]: is there any way we can call these 2 can't from inside the `logs_cache.get` method? 

i think the call would be cleaner this way, but maybe it's just not possible

File: web3/eth.py [hunk 2]
```diff
@@ -57,6 +58,11 @@ class Eth(Module):
     defaultContractFactory = Contract
     iban = Iban
     gasPriceStrategy = None
+    logs_manager: EthLogsManager
```
Comment (mortelli) [inline web3/eth.py]: you're adding a class attribute here. i don't think you want or need this

File: web3/logs_manager.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                for log in log_list:
+                    logs.append(log)
+                current_batch_start = current_batch_end + 1
+                current_batch_end = EthLogsManager.get_next_batch_end(current_batch_start, to_block, batch_size)
+
+        return logs
+
+    @staticmethod
+    def generate_cache_key(address: str, topics: List[str]) -> int:
+        return hash(f"logs_{address}_{topics.sort()}")
```
Comment (mortelli) [inline web3/logs_manager.py]: this one can be removed

File: web3/logs_manager.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                                 f'from block {current_batch_start} to block {current_batch_end}. '
+                                 f'(attempt {currentTry + 1})')
+                    log_list = self.web3.manager.request_blocking(
+                        "eth_getLogs", [filter_params],
+                    )
+                    break
+                except ValueError as e:
+                    logger.warning(f'Error trying to get logs with params {filter_params} '
+                                   f'from block {current_batch_start} to block {current_batch_end}', e)
+            if log_list is None:
```
Comment (mortelli) [inline web3/logs_manager.py]: can `log_list` ever be `None` without entering the exception clause on line `56`?

if not, then we don't need to check for this condition. instead you can raise the `ValueError` after the `logger.warning` call and you don't need to do `log_list = None` on line `46`, which is unpythonic

File: web3/logs_manager.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        filter_params.pop("batchSize", None)
+        filter_params.pop("retries", None)
+
+        current_batch_start: int = int(from_block)
+        current_batch_end: int = EthLogsManager.get_next_batch_end(current_batch_start, to_block, batch_size)
+        logs = []
+        while current_batch_start <= current_batch_end:
+            filter_params["fromBlock"] = current_batch_start
+            filter_params["toBlock"] = current_batch_end
+            log_list = None
```
Comment (mauricioirace) [inline web3/logs_manager.py]: I don't like the name "log_list". But I can't think of a good name either, as "logs" is already defined.

Comment (mauricioirace) [reply web3/logs_manager.py]: Maybe "log_batch" or something explaining it contains the logs of a batch of blocks.

**Review by mauricioirace — APPROVED** [review]
LGTM!

**Review by mortelli — COMMENTED** [review]
most of this seems OK, i've left some comments down below.

i don't like the fact that we have to pass `earliest_block_number` and `latest_block_number` all around. i wish these could be set (or re-set) on the class itself, as this would simplify the signatures for the `get_logs` and `block_identifier_to_number` functions and make the code more readable.

it's not a blocker, but note that both the `from_block` and the `to_block` assignments take 10 lines each. it's too verbose when taking the relatively simple underlying logic into consideration.

i can try to come up with an alternative if you think it's worth the effort. if not, let's just live with it

File: web3/eth.py (hunk 1 above)
Comment (JONAF2103) [reply web3/eth.py:361]: Well i should duplicate code actually, that's why i'm sending this as parameters

File: web3/logs_manager.py (hunk 4 above)
Comment (JONAF2103) [reply web3/logs_manager.py]: You have retries here, if all the retries fail then log_list will be None, and then is when you need to raise, not before that

Comment (mortelli) [reply web3/logs_manager.py]: you're right. my bad!

File: web3/eth.py (hunk 1 above)
Comment (mortelli) [reply web3/eth.py:361]: alright, let's leave it as it is then :+1:

**Review by mortelli — APPROVED** [review]

File: web3/logs_manager.py (hunk 5 above)
Comment (JONAF2103) [reply web3/logs_manager.py]: Done!

File: web3/eth.py (hunk 2 above)
Comment (JONAF2103) [reply web3/eth.py]: Done

**Pull request merged**